src/Build/Construction/Solution/SolutionFile.cs

```diff
@@ -1211,6 +1211,51 @@ internal void ParseNestedProjects()
 
                 proj.ParentProjectGuid = parentProjectGuid;
             } while (true);
+
+            ProjectFileErrorUtilities.VerifyThrowInvalidProjectFile(TestForCyclicalNesting(), "SubCategoryForSolutionParsingErrors",
+                new BuildEventFileInfo(FullPath, _currentLineNumber, 0), "SolutionContainsCyclicallyNestedProjectFiles", _solutionFile);
+        }
+
+        private bool TestForCyclicalNesting()
+        {
+            int[] parents = new int[_projectsInOrder.Count];
+            for (int a = 0; a < _projectsInOrder.Count; a++) {
+                parents[a] = -1;
+            }
+            for (int i = 0; i < _projectsInOrder.Count; i++) {
+                if (!DepthFirstSearch(parents, i))
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        // -2 indicates that there is no cycle among ancestors. -1 indicates that it has not yet been checked.
+        private bool DepthFirstSearch(int[] parents, int index)
```

Comment:
Can we change the name of `DepthFirstSearch` or add a comment here? Maybe `DepthFirstSearch_FindCycle` to add context to this function (particularly about what it means when it returns true or false)